src/main/java/cn/nukkit/utils/BinaryStream.java

```diff
@@ -59,113 +74,221 @@ public void setOffset(int offset) {
         this.offset = offset;
     }
 
+    public byte[][] getBuffers() {
+        if (offset > 0) {
+            byte[] buf2 = new byte[offset];
+            System.arraycopy(buffer, 0, buf2, 0, offset);
+            buffers.addLast(buf2);
+            count += offset;
+            offset = 0;
+        }
+        byte[][] res = new byte[buffers.size()][];
+        int i = 0;
+        for (byte[] bytes : buffers) {
+            res[i++] = bytes;
+        }
+        return res;
+    }
+
     public byte[] getBuffer() {
-        return Arrays.copyOf(buffer, count);
+        if (buffers.size() < 8) {
+            if (buffers.isEmpty()) {
+                buffer = Arrays.copyOfRange(buffer, 0, offset);
+                return buffer;
+            }
+            byte[] data = new byte[getCount()];
+
+            // Check if we have a list of buffers
+            int pos = 0;
+
+            if (buffers != null) {
+                for (byte[] bytes : buffers) {
+                    System.arraycopy(bytes, 0, data, pos, bytes.length);
+                    pos += bytes.length;
+                }
+            }
+
+            // write the internal buffer directly
+            System.arraycopy(buffer, 0, data, pos, offset);
+
+            this.buffer = data;
+            this.buffers.clear();
+            return this.buffer;
+        }
+        final byte[] data = new byte[getCount()];
+        // Check if we have a list of buffers
+        int pos = 0;
+        for (final byte[] bytes : buffers) {
+            final int finalPos = pos;
+            POOL.submit(new Callable() {
```

Comment:
Does an array copy _really_ need a `ForkJoinPool`? I'd imagine this would be slower than just doing it sequentially.